### PR TITLE
Add support for cross directory imports

### PR DESCRIPTION
**Pull request by disconnect3d**
Having a `main.sol`:
```
pragma solidity ^0.4.25;

import '../utils/DirImported.sol';
import 'utils/DirImported2.sol';

contract C {}
```

and such directory structure:
```
$ tree
.
├── main
│   └── main.sol
└── utils
    ├── DirImported.sol
    └── DirImported2.sol

2 directories, 3 files
```

and launching `manticore [options] main/main.sol` Manticore didn't respect current working directory and launched `solc [options] --allow-paths . main.sol` in `main` directory. As a result Manticore was failing due to `solc` being unable to import code from `utils` directory.

This PR fixes that issue and so Manticore will launch `solc [options] --allow-paths . main/main.sol` instead. A testcase has been added for that issue.

---
This change is [Reviewable](https://reviewable.io/reviews/trailofbits/manticore/1233)

### CODE REVIEW

File: tests/test_binaries.py [hunk 1]
```diff
@@ -1,19 +1,34 @@
-import io
 import unittest
 import sys
 import shutil
 import tempfile
 import os
-import hashlib
```
Comment (disconnect3d) [inline tests/test_binaries.py]: Those are redundant imports.

File: tests/test_binaries.py [hunk 2]
```diff
@@ -77,7 +96,7 @@ def _simple_cli_run(self, filename, contract=None, tx_limit=1):
         command.append(str(tx_limit))
         command.append(filename)
 
-        subprocess.check_call(command, stdout=subprocess.PIPE)
+        subprocess.check_call(command, stdout=subprocess.PIPE, cwd=working_dir)
```
Comment (feliam) [inline tests/test_binaries.py]: This is not really checking anything. Just that the process wont take > 10 mins I guess. We should try to add more checks here. Eventually

Comment (disconnect3d) [reply tests/test_binaries.py]: It checks whether Manticore failed its execution or not. Initially I thought that the listed issues had to crash all Manticore and it returns exit code != 0.

But I guess that https://github.com/trailofbits/manticore/issues/676 did return 0 exit code, so I am confused a bit.

**Comment by disconnect3d** [comment]
One thing I am worried about is this code:
https://github.com/trailofbits/manticore/blob/master/manticore/ethereum/__init__.py#L280-L290

If the solidity contract path is very long, the `lib_placeholder` in:
```python
lib_name = lib_placeholder.split(':')[1].split('_')[0]
```

will end up not having a `:` in it as apparently `solc` truncates the path to a given number of characters... so `[1]` will end up in a `IndexError`.

This was thrown in one of the detector testcases that used `library {...}` - https://github.com/trailofbits/manticore/blob/master/tests/binaries/detectors/delegatecall_ok.sol

I mitigated it by adding a `working_dir` to the `solidity_create_contract` call. But in the end, if one doesn't pass a `working_dir` this may still fail.

Anyway, I think the `solidity_create_contract` should still have a possibility of specifying working directory, so imports will work fine in contracts used through the API.

File: manticore/ethereum/__init__.py [hunk 3]
```diff
@@ -336,27 +336,33 @@ def _run_solc(source_file, solc_bin=None, solc_remaps=[]):
             #logger.warning("Unsupported solc version %s", installed_version)
             pass
 
-        #shorten the path size so library placeholders wont fail.
```
Comment (feliam) [inline manticore/ethereum/__init__.py]: Can notwe use this path shortening thing in `solidity_create_contract` if the working_directory is not specified?

File: manticore/ethereum/__init__.py [hunk 4]
```diff
@@ -336,27 +336,33 @@ def _run_solc(source_file, solc_bin=None, solc_remaps=[]):
...
-        p = Popen(solc_invocation, stdout=PIPE, stderr=PIPE, cwd=working_folder)
+        p = Popen(solc_invocation, stdout=PIPE, stderr=PIPE, **additional_kwargs)
         stdout, stderr = p.communicate()
 
         stdout, stderr = stdout.decode(), stderr.decode()
 
         # See #1123 - solc fails when run within snap
         # and https://forum.snapcraft.io/t/interfaces-allow-access-tmp-directory/5129
-        if stdout == '' and '""%s"" is not found' % filename in stderr:
+        if stdout == '' and '""%s"" is not found' % relative_filepath in stderr:
```
Comment (feliam) [inline manticore/ethereum/__init__.py]: Maybe just f-string it?

File: manticore/ethereum/__init__.py (hunk 3 above)
Comment (disconnect3d) [reply manticore/ethereum/__init__.py]: We resolved this on a DM; for others/future ref:

if someone lanuches:
```manticore some_path```
it should launch `solc` with the same path and from the same working dir as the person is on -
unless the person wants it differently.

Previously we shortened the path but we can't do it as we can't know how much can we shorten it as there might be relative imports in contracts :s.

**Pull request merged**